extensions/functions_string.yaml

```diff
@@ -251,6 +251,104 @@ scalar_functions:
             name: "replacement"
             description: The replacement string.
         return: "varchar<L1>"
+  -
+    name: regexp_replace
+    description: >-
+      Search a string for a substring that matches a given regular expression pattern and replace
+      it with a replacement string. The regular expression pattern should follow the
+      International Components for Unicode implementation (https://unicode-org.github
+      .io/icu/userguide/strings/regexp.html). The occurrence of the pattern to be replaced is
+      specified using the `occurrence` argument. Specifying `1` means only the first occurrence
+      will be replaced and specifying `0` means all occurrences will be replaced. The number of
+      characters from the beginning of the string to begin starting to search for pattern matches
+      can be specified using the `position` argument.  Specifying `1` means to search for matches
+      starting at the first character of the input string. The replacement string can capture
+      groups using numbered backreferences.
```

Comment:
It'd be nice to be more explicit about how > 1 works, e.g. "Specifying `1` means only the first occurrence will be replaced, `2` means the second occurrence, and so on. Specify `0` to replace all occurrences." and likewise for the string indices.

What happens if I were to pass negative numbers or 0 for the string index?

Reply:
Updated.  For string index, it states that it should be a positive non-zero integer now.  How does that sound?

Reply:
I already understood that I'm not supposed to do it, but it still doesn't really specify what would happen if I would :) The same actually applies for what happens if the regex fails to compile, or the replacement has a back-reference that doesn't exist. Maybe this is overly pedantic and it should be up to the engine to decide whether to invoke `MAYBE_HALT_AND_CATCH_FIRE`, throw a segfault, or do something more gentle, but otherwise I see three options:

 - throw an error and fail the entire plan if any invocation fails;
 - change nullability behavior to `SPECIFIED_OUTPUT` and always return nullable, so you can fail invocations independently by returning null;
 - allow either of the above by means of an option enum.

Reply:
Yeah, I was originally under the assumption that the engine would decide, which I guess would open up a lot of variability between engine implementations. Do you have a preference for any of those three options?

Reply:
Seems like not that many actually have this as an argument that can be passed.  MySQL does though and if you pass a negative input, it throws and error saying incorrect parameters.

Reply:
> How do you suggest we define the option that allows 1 or 2? [ DECLARED_OUTPUT, ERROR]?

I was thinking something like an option named "on_error" with values `ERROR` and `NULL`.

> And would we need to explicitly set the nullability for the implementations?

Yes, and that's actually the primary reason I'm kind of on the fence about it. There's no way to specify that the function should have MIRROR behavior for `ERROR` and `DECLARED_OUTPUT` behavior for `NULL`, so you'd have to do `DECLARED_OUTPUT` for both. That's kind of nasty and will probably not work well with the functions in existing query engines.

I may well be overthinking it. Maybe we should just add something like "Behavior is undefined if the regex fails to compile, the replacement contains an illegal back-reference, the occurrence value is out of range, or the position value is out of range." to leave it up to the engine. Or just keep things as is and not specify anything, since that statement is arguably redundant (though I'd prefer to be overly explicit to show that we did actually think about these edge cases).

I guess I'll leave this up to you and @jacques-n (or anyone else stumbling upon this who might have better ideas).

Reply:
I went with adding your error handling message you wrote there to the description.  I'm also on the fence about potentially having this nullability situation here.